gatsby/content/blog/2021/12/2021-12-10-sydent-typing-2.mdx

```diff
@@ -0,0 +1,384 @@
+---
+  date: '2021-12-10'
+  title: 'Type coverage for Sydent: annotation'
+  categories:
+    - Tech
+  author: David Robertson
+---
+
+This is the second in a series of three posts which discuss recent work to improve type annotations in [Sydent](https://github.com/matrix-org/sydent), the reference Matrix Identity server. [Last time](2021-12-03-sydent-typing-1.mdx) we discussed the motivation for doing this work in the first place: the _why_. Now I want to talk about the _how_. How did we add annotations to individual files, and across the project as whole? What common idioms did we learn on the way?
+
+## The process of improving coverage
+
+
+From experience adding typing to Synapse, we decided to annotate one module at a time. We configured a list of files in `pyproject.toml` which we knew passed mypy's checks. We could run `mypy` in CI to check that new PRs didn't introduce type problems in modules already covered.
+
+From there, the workflow was
+
+1. Choose a new module to annotate. Add it to the list of `files` in mypy's configuration.
+2. Run `mypy`. See how many errors you get.
+3. Choose an error. Fix it. Re-run `mypy`.
+4. Repeat until no errors remaining.
+5. Submit for review.
+
+There are two parts in there which involve a choice. Being honest, I made those choices unscientifically: I tried to choose the easy tasks to do first. My first target was actually the entire [`sydent.util`](https://github.com/matrix-org/sydent/pull/418) subpackage. Probably a bit too large for a first bite! My thinking was that `util` sounded like something with few dependencies that would have impact across the whole source tree.
+
+Within a given module, I'd try to fix easier errors first: partly for confidence, partly to build up momentum, and partly to get myself familiar with that piece of source code. For example, I'd often start by telling mypy that `mylist = []` was actually was a `List[str]` (rather than the generic `List[Any]` which it would use otherwise).
+
+Picking and choosing easy targets works fairly well, but sometimes that means you end fixing an error that's really a symptom of an earlier one. Other times fixing one error, e.g. by giving a return type annotation to a function—would solve a series of other errors throughout the file. Watching the total number of errors mypy reports bob up and down was intriguing!
+
+In retrospect, I think it would be smoother to generate some kind of dependency graph for the package. I'm imagining a [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) where whose vertices are modules, and there's an edge `A -> B` if `A` imports from `B`. The leaves of this DAG are the ideal place to start: you can get something typechecked there without having to annotate a long chain of dependencies. Another strategy would be to see which modules were the least precise according to mypy's reports—but more on those [next time](2021-12-03-sydent-typing-1.mdx).
```

Comment:
The "next time" link 404's

Reply:
and it actually links to the previous time!

Reply:
Should be fixed in a7185f82